File: Data Structures & Algorithms/baseball-game/submission-3.cpp
#include <string>

class Solution {
public:
    int calPoints(vector<string>& operations) {
        int score = 0;
        stack<int> s;
        for(int i=0; i<operations.size(); i++){
            
            if(operations[i]=="D"){
                s.push(s.top()*2);
            }
            else if(operations[i]=="+"){
                int first = s.top(); 
                s.pop();

                int temp = s.top() + first;
                s.push(first);
                s.push(temp);
            }
            else if(operations[i]=="C"){
                s.pop();
            }
            else{
                s.push(stoi(operations[i]));
            }
        }
        while (!s.empty()) {
            score += s.top();
            s.pop();
        }
        return score;
    }
};

File: Data Structures & Algorithms/minimum-stack/submission-3.py
class MinStack:

    def __init__(self):
        self.s1 = []
        self.s2 = []

    def push(self, val: int) -> None:
        self.s1.append(val)
        if self.s2 and val<self.s2[-1]:
            self.s2.append(val)
        elif self.s2:
            self.s2.append(self.s2[-1])
        else:
            self.s2.append(val)

    def pop(self) -> None:
        self.s1.pop()
        self.s2.pop()
        
    def top(self) -> int:
        return self.s1[-1]

    def getMin(self) -> int:
        return self.s2[-1]
        
        
